### ISSUE TITLE
[Table] When scrolling, lock to columns and rows to improve perf

### ISSUE THREAD

**Issue by cmslewis** [palantir/blueprint#887]
`Table` scrolling currently feels choppy and unpolished when the viewport is large (see #390):
- Column and row headers get misaligned with their respective body cells,
- the body flashes white as it tries to catch up to fast scrolls to a distant vantage point, and
- the frame rate drops significantly below a smooth 60fps (I've seen it drop to as low as 4fps).

One way to increase frames per second is to decrease the number of frames to render.

Here's what we currently do:
![2017-03-24 15 55 57](https://cloud.githubusercontent.com/assets/443450/24316574/41b371c6-10ac-11e7-9046-d1b8daae2aca.gif)

In contrast, here's what Google Sheets does:
![2017-03-24 16 07 17](https://cloud.githubusercontent.com/assets/443450/24316542/010c15e2-10ac-11e7-8ffd-a41ab1b0aaa1.gif)

Whereas our `Table` scrolls smoothly _between_ cell boundaries, Google Sheets scrolls _in lockstep_ with cell boundaries. This actually makes a lot of sense for two reasons:

1. From a **UX** perspective, a half-visible cell is hardly more useful than a non-visible cell. Scrolling directly to the next row or column ensures that only complete cells are visible.
2. From a **performance** perspective, the lockstep approach eliminates intermediate frames between cell boundaries. Thus, if I scroll vertically through 10 rows, that amounts to `10` frames, not `10 * ~60 frames`, a huge savings.

Although Issue #390 deals with the same thematic problem as this new proposal, I opened this new issue to track the specific task of improving `Table`'s scrolling behavior and nothing more. There may still be more work required before we can close #390 for good.

**Comment by leebyp** [comment]
An edge case but one thing to note is that it with a lock, it becomes impossible to see the tail of a cell if the column is wider than the viewport. That being said, I'd take a 60x perf improvement any day =p

**Comment by cmslewis** [comment]
Investigated this for awhile. Managed to make a JSFiddle successfully demo'ing this behavior with a native HTML `<table>`, but I got into a real mess when I tried to get it working on our Blueprint `Table`. Moreover, the approach in my JSFiddle had a pretty glaring usability issue: it requires an invisible scrollable overlay that swallows all mouse interactions.

Zooming out, profiling reveals that perf issues in general are mostly attributable to React internal functions. The solution to our woes will probably be to abandon React lifecycle methods internally within the `Table` component and handle state changes manually. // CC @bdwyer, who's suggested this before

Going to move on to other Table issues for now.

**Comment by adidahiya** [comment]
> The solution to our woes will probably be to abandon React lifecycle methods internally within the Table component and handle state changes manually.

Are you sure we're not spending too many cycles doing unnecessary work in these lifecycle methods? Not sure we need to jump ship quite yet.

**Comment by cmslewis** [comment]
Not completely sure, no. But going through the profiler in Chrome, it seems very little of the scripting time between paints is spent inside functions our components provide.

**Comment by cmslewis** [comment]
Here's the JSFiddle with prototype functionality: https://jsfiddle.net/cmslewis/m4dq8pyj/22/. Note that there's a giant scroll area DIV covered the table elements though, so you can't really access the cells directly.

**Comment by adidahiya** [comment]
@cmslewis have you tried `pointer-events: none`? It works in our [supported browsers](http://caniuse.com/#search=pointer-events).

**Comment by cmslewis** [comment]
Does that capture scroll events (e.g. trackpad moves) though?

On Apr 4, 2017, at 4:38 PM, Adi Dahiya <notifications@github.com<mailto:notifications@github.com>> wrote:


@cmslewis[github.com]<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_cmslewis&d=DwMCaQ&c=izlc9mHr637UR4lpLEZLFFS3Vn2UXBrZ4tFb6oOnmz8&r=-xHM_msPNZ5Ueek6I-8Uf3C2NKIdRgHQEGOCuANP978&m=uf-BDsGvf6l0rwq1TqIvtmo9H0ZQyhVpL_ipa7AT9Zs&s=zWrGDBIPYmVQei7tkrTy-pjT8IPI4VyDjIVuMmwXZS4&e=> have you tried pointer-events: none? It works in our supported browsers[caniuse.com]<https://urldefense.proofpoint.com/v2/url?u=http-3A__caniuse.com_-23search-3Dpointer-2Devents&d=DwMCaQ&c=izlc9mHr637UR4lpLEZLFFS3Vn2UXBrZ4tFb6oOnmz8&r=-xHM_msPNZ5Ueek6I-8Uf3C2NKIdRgHQEGOCuANP978&m=uf-BDsGvf6l0rwq1TqIvtmo9H0ZQyhVpL_ipa7AT9Zs&s=7hKrvFuulWGAoyHpsiKnpJcAbvOMTahCNAxZLUejxY8&e=>.

**Comment by adidahiya** [comment]
ah, you're right, it doesn't solve the issue. nevermind.

**Comment by cmslewis** [comment]
@tgreenwatts think we can close this one? We're solving perf in other ways, and the scrolling-UX change here is questionable.
